docs/reference/components/sources/internal_metrics.cue

```diff
@@ -282,6 +282,12 @@ components: sources: internal_metrics: {
 			default_namespace: "vector"
 			tags:              _component_tags
 		}
+		kafka_statistics_received_total: {
+			description:       "The total number of statistics that have been received from Kafka."
+			type:              "counter"
+			default_namespace: "vector"
+			tags:              _component_tags
+		}
```

Comment:
I'm confused, shouldn't we be emitting the actual Kafka statistics, not just a count of them?

Reply:
This is a workaround I chose, and definitely up to debate. The [`metrics!`](https://docs.rs/metrics/0.13.0-alpha.13/metrics/index.html) facade only provides a couple of methods to emit numeric values.

Therefore, I simply emit a counter for the number of statistics emitted, and provide the actual statistics JSON in the `statistics={...}` label along the event.

Reply:
So, at the moment it looks like this:

```
2021-01-10T10:09:16.843781Z vector_kafka_statistics{component_kind="source",component_name="kafka",component_type="kafka",statistics="{...}"} = 1
```

Ideally, we would want


```
2021-01-10T10:09:16.843781Z vector_kafka_statistics{component_kind="source",component_name="kafka",component_type="kafka"} = {...}
```

correct?

@MOZGIII (since I believe you added the `metrics!` implementation), do you have some insight into how I can emit a non-numeric event through the same hooks that the `metrics!` go through?

Reply:
Hm, that's very strange. I think there might be a fundamental misunderstanding. Why wouldn't we emit each relevant Kafka statistic as it's own metric, just like any other metric?

Reply:
That was my initial question in this PR:

> Do we want to break up / flatten the statistics provided by rdkafka and emit an event for each of the properties (instead of one JSON string)?

Unfortunately my wording might not have stated the problem well enough? 😅 I can go back to that approach, though.

Reply:
Yes, we'll need to do that and adjust the names of the metrics to match our internal metric naming conventions. We should not blindly pass through every single statistic, but instead, represent them as our own statistics. That means white-listing statistics that are:

1. Obviously valuable.
2. Can be represented as a metric.

In regards to string information, use your best judgment if you think that should be represented as a metric tag. I would default to not including it until we identify a need for it.

Reply:
> Why wouldn't we emit each relevant Kafka statistic as it's own metric, just like any other metric?

There are relatively many of them, and these can be nested up to five levels (e.g. `topics.$topic.partitions.$partition.txmsgs`). Looking at it from the user side, I wasn't sure how much value there is in splitting these up into one event type for each of them, instead of extracting the desired information from the JSON provided by Kafka directly.

Also, these access paths would be heavily duplicated. E.g. for each property of the `partition`, we would log

```
kafka_partition_topic_partition_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_broker_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_leader_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_desired_...{...,topic=$topic,partition=$partition}: false,
kafka_partition_topic_unknown_...{...,topic=$topic,partition=$partition}: false,
kafka_partition_topic_msgq_cnt_...{...,topic=$topic,partition=$partition}: 845,
kafka_partition_topic_msgq_bytes_...{...,topic=$topic,partition=$partition}: 26195,
kafka_partition_topic_xmit_msgq_cnt_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_xmit_msgq_bytes_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_fetchq_cnt_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_fetchq_size_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_fetch_state_...{...,topic=$topic,partition=$partition}: "none",
kafka_partition_topic_query_offset_...{...,topic=$topic,partition=$partition}: -1001,
kafka_partition_topic_next_offset_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_app_offset_...{...,topic=$topic,partition=$partition}: -1001,
kafka_partition_topic_stored_offset_...{...,topic=$topic,partition=$partition}: -1001,
kafka_partition_topic_committed_offset_...{...,topic=$topic,partition=$partition}: -1001,
kafka_partition_topic_eof_offset_...{...,topic=$topic,partition=$partition}: -1001,
kafka_partition_topic_lo_offset_...{...,topic=$topic,partition=$partition}: -1001,
kafka_partition_topic_hi_offset_...{...,topic=$topic,partition=$partition}: -1001,
kafka_partition_topic_ls_offset_...{...,topic=$topic,partition=$partition}: -1001,
kafka_partition_topic_consumer_lag_...{...,topic=$topic,partition=$partition}: -1,
kafka_partition_topic_txmsgs_...{...,topic=$topic,partition=$partition}: 3950967,
kafka_partition_topic_txbytes_...{...,topic=$topic,partition=$partition}: 122479977,
kafka_partition_topic_rxmsgs_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_rxbytes_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_msgs_...{...,topic=$topic,partition=$partition}: 3951812,
kafka_partition_topic_rx_ver_drops_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_msgs_inflight_...{...,topic=$topic,partition=$partition}: 1067,
kafka_partition_topic_next_ack_seq_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_next_err_seq_...{...,topic=$topic,partition=$partition}: 0,
kafka_partition_topic_acked_msgid_...{...,topic=$topic,partition=$partition}: 0
```

instead of logging

```
{
    ...,
    "partitions": {
        ...,
        "0": {
            "partition": 0,
            "broker": 0,
            "leader": 0,
            "desired": false,
            "unknown": false,
            "msgq_cnt": 845,
            "msgq_bytes": 26195,
            "xmit_msgq_cnt": 0,
            "xmit_msgq_bytes": 0,
            "fetchq_cnt": 0,
            "fetchq_size": 0,
            "fetch_state": "none",
            "query_offset": -1001,
            "next_offset": 0,
            "app_offset": -1001,
            "stored_offset": -1001,
            "committed_offset": -1001,
            "eof_offset": -1001,
            "lo_offset": -1001,
            "hi_offset": -1001,
            "ls_offset": -1001,
            "consumer_lag": -1,
            "txmsgs": 3950967,
            "txbytes": 122479977,
            "rxmsgs": 0,
            "rxbytes": 0,
            "msgs": 3951812,
            "rx_ver_drops": 0,
            "msgs_inflight": 1067,
            "next_ack_seq": 0,
            "next_err_seq": 0,
            "acked_msgid": 0
        },
        ...
    },
    ...
}
```

once somewhere within that JSON.

Reply:
These statistics from the [top-level](https://github.com/edenhill/librdkafka/blob/master/STATISTICS.md#top-level) look like the highest-value ones to me:

- `msg_cnt` - Current number of messages in producer queues
- `msg_size` - Current total size of messages in producer queues
- `tx` - Total number of requests sent to Kafka brokers
- `tx_bytes` - Total number of bytes transmitted to Kafka brokers
- `rx` - Total number of responses received from Kafka brokers
- `rx_bytes` - Total number of bytes received from Kafka brokers
- `txmsgs` - Total number of messages transmitted (produced) to Kafka brokers
- `txmsg_bytes` - Total number of message bytes (including framing, such as per-Message framing and MessageSet/batch framing) transmitted to Kafka brokers
- `rxmsgs` - Total number of messages transmitted (produced) to Kafka brokers
- `rxmsg_bytes` - Total number of message bytes (including framing) received from Kafka brokers

If there will be user demand for more specific ones, we can add them in the future, I suppose.

Reply:
That list looks good to me. Agree, we can add more based on demand. To reiterate, we follow the [Prometheus style](https://prometheus.io/docs/practices/naming/) naming. So the actual metric names are fairly broad, leveraging tags for dimensions. Things like the partition name, etc, should not be in the actual metric name.